swaybg/main.c

```diff
@@ -21,6 +22,99 @@ enum scaling_mode {
 	SCALING_MODE_TILE,
 };
 
+
+#ifndef GDK_PIXBUF_CHECK_VERSION
+#define GDK_PIXBUF_CHECK_VERSION(major,minor,micro) \
+		(GDK_PIXBUF_MAJOR > (major) || \
+		 (GDK_PIXBUF_MAJOR == (major) && GDK_PIXBUF_MINOR > (minor)) || \
+		 (GDK_PIXBUF_MAJOR == (major) && GDK_PIXBUF_MINOR == (minor) && \
+			GDK_PIXBUF_MICRO >= (micro)))
+#endif
+
+cairo_surface_t *
+__gdk_pixbuf_create_cairo_surface(const GdkPixbuf *gdkbuf)
+{
+
+	int chan = gdk_pixbuf_get_n_channels(gdkbuf);
+	if (chan < 3) return NULL;
+
+	cairo_format_t fmt = (chan == 3) ? CAIRO_FORMAT_RGB24 : CAIRO_FORMAT_ARGB32;
+
+#if GDK_PIXBUF_CHECK_VERSION(2,32,0)
+	const guint8* gdkpix = gdk_pixbuf_read_pixels(gdkbuf);
+#else
+	const guint8* gdkpix = gdk_pixbuf_get_pixels(gdkbuf);
+#endif
+
+	gint w = gdk_pixbuf_get_width(gdkbuf);
+	gint h = gdk_pixbuf_get_height(gdkbuf);
+	int stride = gdk_pixbuf_get_rowstride(gdkbuf);
+	
+	
+	cairo_surface_t * cs = cairo_image_surface_create (fmt, w, h);
+	cairo_surface_flush (cs);
+	if ( !cs || cairo_surface_status (cs) != CAIRO_STATUS_SUCCESS)
+		return NULL;
+
+	int cstride = cairo_image_surface_get_stride (cs);
+	unsigned char * cpix = cairo_image_surface_get_data (cs);
+
+	if (chan == 3) {
+		int i;
+		for (i = h; i; --i) {
+			const guint8 *gp = gdkpix;
+			unsigned char *cp = cpix;
+			const guint8* end = gp + 3*w;
+			while (gp < end) {
+#if G_BYTE_ORDER == G_LITTLE_ENDIAN
+				cp[0] = gp[2];
+				cp[1] = gp[1];
+				cp[2] = gp[0];
+#else
+				cp[1] = gp[0];
+				cp[2] = gp[1];
+				cp[3] = gp[2];
+#endif
+				gp += 3;
+				cp += 4;
+			}
+			gdkpix += stride;
+			cpix += cstride;
+		}
+	} else {
+#define FIXPMUL(x,a,b) G_STMT_START { guint t = a * b + 0x80; x = (t + (t >> 8)) >> 8; } G_STMT_END
+		// premultiply alpha as fixed point 8.8
+		// [(t + 1 + ( (t + 1) >> 8)) >> 8] is approximately t div 255
+		// replaced 1 with 0x80 for rounding purposes
```

Comment:
I think I should remove the comment about fixed point math. That is no longer true, since colors work with 255==1 instead of 256==1.
I need good results for a very specific range, numbers from 0 to 255 and there it works very well. (the linear projection of [0..1] -> [0..255]). ```
def premul(primary_color, alpha)
    return round(primary_color/255 * alpha/255 * 255)
premul(0xff, 0xff) = 0xff
premul(0xff, 0x80) = 0x80 (127.5 rounded up)
premul(0x80, 0xff) = 0x80
```
In my python script it works very well:
```
for x in range(0,256):
    for a in range(0,256):
        t=x*a+0x80
        bs=((t>>8)+t)>>8
        r=round(x*a/255)
        if r != bs:
            print(x, a, r, bs)
```